File: Algorithms/src/1622. Fancy Sequence/Code/solution.java
class Fancy {

    static final long MOD = 1000000007;

    ArrayList<Long> seq = new ArrayList<>();
    long mul = 1;
    long add = 0;

    private long modPow(long a, long b) {
        long res = 1;
        while (b > 0) {
            if ((b & 1) == 1)
                res = (res * a) % MOD;
            a = (a * a) % MOD;
            b >>= 1;
        }
        return res;
    }

    public Fancy() {
    }

    public void append(int val) {
        long inv = modPow(mul, MOD - 2);
        long stored = ((val - add + MOD) % MOD * inv) % MOD;
        seq.add(stored);
    }

    public void addAll(int inc) {
        add = (add + inc) % MOD;
    }

    public void multAll(int m) {
        mul = (mul * m) % MOD;
        add = (add * m) % MOD;
    }

    public int getIndex(int idx) {
        if (idx >= seq.size())
            return -1;
        return (int) ((seq.get(idx) * mul % MOD + add) % MOD);
    }
}

File: Algorithms/src/1622. Fancy Sequence/Code/solution.go
const MOD int64 = 1e9 + 7

type Fancy struct {
    seq []int64
    mul int64
    add int64
}

func modPow(a, b int64) int64 {
    res := int64(1)
    for b > 0 {
        if b&1 == 1 {
            res = res * a % MOD
        }
        a = a * a % MOD
        b >>= 1
    }
    return res
}

func Constructor() Fancy {
    return Fancy{[]int64{}, 1, 0}
}

func (this *Fancy) Append(val int) {
    inv := modPow(this.mul, MOD-2)
    stored := ((int64(val)-this.add+MOD)%MOD * inv) % MOD
    this.seq = append(this.seq, stored)
}

func (this *Fancy) AddAll(inc int) {
    this.add = (this.add + int64(inc)) % MOD
}

func (this *Fancy) MultAll(m int) {
    this.mul = this.mul * int64(m) % MOD
    this.add = this.add * int64(m) % MOD
}

func (this *Fancy) GetIndex(idx int) int {
    if idx >= len(this.seq) {
        return -1
    }
    return int((this.seq[idx]*this.mul%MOD + this.add) % MOD)
}